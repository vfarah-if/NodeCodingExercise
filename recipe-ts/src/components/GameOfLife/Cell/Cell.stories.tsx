import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';
import Cell, { CellProps } from './Cell';

export default {
  title: 'Component/GameOfLife/Cell',
  component: Cell,
} as Meta;

const CellTemplate: Story<CellProps> = (args) => (
  <>
    <Cell {...args} />
  </>
);

export const DefaultCell = CellTemplate.bind({});
DefaultCell.args = {};
