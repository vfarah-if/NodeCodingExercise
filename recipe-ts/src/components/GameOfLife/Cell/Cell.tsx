import React, { MouseEventHandler } from 'react';
import './style/index.css';

export interface CellProps {
  cellSize?: number;
  cellBorderColor?: string;
  cellBackgroundColor?: string;
  cellActiveColor?: string;
  displayText?: string;
  isActive?: boolean;

  onCellClick?: MouseEventHandler;
}

const Cell: React.FC<CellProps> = ({
  cellSize = 40,
  cellBorderColor = 'gray',
  cellBackgroundColor,
  cellActiveColor = 'blue',
  displayText,
  isActive = false,
  onCellClick,
}) => {
  return (
    <div
      className={isActive ? 'cell active' : 'cell'}
      onClick={onCellClick}
      style={{
        width: cellSize,
        height: cellSize,
        border: `solid 1px ${cellBorderColor}`,
        backgroundColor: `${isActive ? cellActiveColor : cellBackgroundColor}`,
      }}
    >
      {displayText}
    </div>
  );
};

export default Cell;
