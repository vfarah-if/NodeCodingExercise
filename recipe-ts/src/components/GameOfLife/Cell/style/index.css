.cell {
  width: 20;
  height: 20;
  border: solid 1px black;
  text-align: center;
}
