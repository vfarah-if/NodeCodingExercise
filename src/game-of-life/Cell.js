const { CellState } = require('./CellState');

class Cell {
	constructor(state = CellState.Dead) {
		this.currentState = state;
		this.neighbours = new Array();
	}

	addNeighbours(items) {
		items?.forEach((item) => {
			this.neighbours.push(item);
		});
	}

	nextState() {
		let result = this.currentState;
		const liveNeighbours = this.neighbours.filter(
			(item) => item.currentState === CellState.Alive
		);
		const isFertile =
			liveNeighbours.length === 3 && this.currentState === CellState.Dead;
		if (isFertile) {
			result = CellState.Alive;
		}

		const isOverpopulated = liveNeighbours.length > 3;
		const isUnderpopulated = liveNeighbours.length < 2;
		if (isUnderpopulated || isOverpopulated) {
			result = CellState.Dead;
		}
		return result;
	}
}

module.exports = Cell;
