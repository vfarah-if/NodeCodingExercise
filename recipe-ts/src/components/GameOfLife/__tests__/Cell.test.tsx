import { render } from '@testing-library/react';
import { BrightHappyCell, DefaultCell } from '../Cell/Cell.stories';

describe('Cell', () => {
  test('should render a cell with default values', () => {
    const { container } = render(<DefaultCell {...DefaultCell.args} />);

    expect(container).toMatchSnapshot();
  });

  test('should render a cell with configured values', () => {
    const { container } = render(<BrightHappyCell {...BrightHappyCell.args} />);

    expect(container).toMatchSnapshot();
  });
});
