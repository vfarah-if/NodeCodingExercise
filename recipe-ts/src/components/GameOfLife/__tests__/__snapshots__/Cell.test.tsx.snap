// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell should render a cell with configured values 1`] = `
<div>
  <div
    class="cell"
    style="width: 50px; height: 50px; border: 1px solid orange; background-color: yellow;"
  >
    (ãƒ„)
  </div>
</div>
`;

exports[`Cell should render a cell with default values 1`] = `
<div>
  <div
    class="cell"
    style="width: 40px; height: 40px; border: 1px solid gray;"
  />
</div>
`;
